compilation error when sasl.h is missing

Error compiling memcached extension even having sasl disabled in configure.
 
```
# pecl download redis memcached
# cd memcached-3.0.3/
# phpize
 # ./configure --disable-memcached-sasl
...
checking for libmemcached... yes, shared
checking whether to enable memcached session handler support... yes
checking whether to enable memcached igbinary serializer support... no
checking whether to enable memcached json serializer support... no
checking whether to enable memcached msgpack serializer support... no
checking whether to enable memcached sasl support... no
checking whether to enable memcached protocol support... no
...
creating libtool
appending configuration tag "CXX" to libtool
configure: creating ./config.status
config.status: creating config.h
# make
...
 cc -I/usr/local/include/php -I. -I/usr/local/memcached-3.0.3 -DPHP_ATOM_INC -I/usr/local/memcached-3.0.3/include -I/usr/local/memcached-3.0.3/main -I/usr/local/memcached-3.0.3 -I/usr/local/include/php -I/usr/local/include/php/main -I/usr/local/include/php/TSRM -I/usr/local/include/php/Zend -I/usr/local/include/php/ext -I/usr/local/include/php/ext/date/lib -DHAVE_CONFIG_H -g -O2 -I/usr/local/include/php -I/usr/local/include/php/main -I/usr/local/include/php/TSRM -I/usr/local/include/php/Zend -I/usr/local/include/php/ext -I/usr/local/include/php/ext/date/lib -c /usr/local/memcached-3.0.3/php_memcached.c  -fPIC -DPIC -o .libs/php_memcached.o
In file included from /usr/include/libmemcached-1.0/memcached.h:85:0,
                 from /usr/include/libmemcached/memcached.h:39,
                 from /usr/local/memcached-3.0.3/php_libmemcached_compat.h:21,
                 from /usr/local/memcached-3.0.3/php_memcached_private.h:28,
                 from /usr/local/memcached-3.0.3/php_memcached.c:25:
/usr/include/libmemcached-1.0/struct/sasl.h:39:23: fatal error: sasl/sasl.h: No such file or directory
compilation terminated.
Makefile:194: recipe for target 'php_memcached.lo' failed
make: *** [php_memcached.lo] Error 1
```

For some reason there is an unconditional #include <libmemcached-1.0/struct/sasl.h> in memcached.h

OS: Alpine linux 3.4.6

--- a/config.m4
+++ b/config.m4
@@ -184,7 +184,7 @@ if test "$PHP_MEMCACHED" != "no"; then
     else
       AC_MSG_RESULT([$msgpack_inc_path])
     fi
-  fi 
+  fi
 
   AC_MSG_CHECKING([for memcached session support])
   if test "$PHP_MEMCACHED_SESSION" != "no"; then
@@ -255,42 +255,45 @@ if test "$PHP_MEMCACHED" != "no"; then
     PHP_EVAL_LIBLINE($PHP_LIBMEMCACHED_LIBS, MEMCACHED_SHARED_LIBADD)
     PHP_EVAL_INCLINE($PHP_LIBMEMCACHED_INCLUDES)
 
-    #
-    # Added -lpthread here because AC_TRY_LINK tests on CentOS 6 seem to fail with undefined reference to pthread_once
-    #
+    dnl # Added -lpthread here because AC_TRY_LINK tests on CentOS 6 seem to fail with undefined reference to pthread_once
     ORIG_CFLAGS="$CFLAGS"
     CFLAGS="$CFLAGS $INCLUDES"
 
+    dnl # Always check if libmemcached was built with SASL support,
+    dnl # because it will require sasl.h even if not used here.
+    AC_CACHE_CHECK([for libmemcached sasl.h requirement], ac_cv_memc_sasl_support, [
+      AC_TRY_COMPILE(
+        [ #include <libmemcached/memcached.h> ],
+        [
+        #if LIBMEMCACHED_WITH_SASL_SUPPORT
+          /* yes */
+        #else
+        #  error "no sasl support"
+        #endif
+        ],
+        [ ac_cv_memc_sasl_support="yes" ],
+        [ ac_cv_memc_sasl_support="no" ]
+      )
+    ])
+
+    if test "$ac_cv_memc_sasl_support" = "yes"; then
+      AC_CHECK_HEADERS([sasl/sasl.h], [ac_cv_have_memc_sasl_h="yes"], [ac_cv_have_memc_sasl_h="no"])
+    fi
+
+    dnl # If libmemcached requires sasl.h but we can't find sasl.h, that's a hard error
+    dnl # regardless of the option --enable-memcached-sasl or --disable-memcached-sasl
     AC_MSG_CHECKING([whether to enable sasl support])
+    if test "$ac_cv_memc_sasl_support" = "yes" && test "$ac_cv_have_memc_sasl_h" = "no"; then
+      AC_MSG_ERROR([no, libmemcached built with sasl required, but sasl.h not found.])
+    fi
+
     if test "$PHP_MEMCACHED_SASL" != "no"; then
       AC_MSG_RESULT(yes)
-      AC_CHECK_HEADERS([sasl/sasl.h], [ac_cv_have_memc_sasl_h="yes"], [ac_cv_have_memc_sasl_h="no"])
-
-      if test "$ac_cv_have_memc_sasl_h" = "yes"; then
-
-        AC_CACHE_CHECK([whether libmemcached supports sasl], ac_cv_memc_sasl_support, [
-          AC_TRY_COMPILE(
-            [ #include <libmemcached/memcached.h> ],
-            [ 
-            #if LIBMEMCACHED_WITH_SASL_SUPPORT
-              /* yes */
-            #else
-            #  error "no sasl support"
-            #endif
-            ],
-            [ ac_cv_memc_sasl_support="yes" ],
-            [ ac_cv_memc_sasl_support="no" ]
-          )
-        ])
-
-        if test "$ac_cv_memc_sasl_support" = "yes"; then
-          PHP_CHECK_LIBRARY(sasl2, sasl_client_init, [PHP_ADD_LIBRARY(sasl2, 1, MEMCACHED_SHARED_LIBADD)])
-          AC_DEFINE(HAVE_MEMCACHED_SASL, 1, [Have SASL support])
-        else
-          AC_MSG_ERROR([no, libmemcached sasl support is not enabled. Run configure with --disable-memcached-sasl to disable this check])
-        fi
+      if test "$ac_cv_memc_sasl_support" = "yes" && test "$ac_cv_have_memc_sasl_h" = "yes"; then
+        PHP_CHECK_LIBRARY(sasl2, sasl_client_init, [PHP_ADD_LIBRARY(sasl2, 1, MEMCACHED_SHARED_LIBADD)])
+        AC_DEFINE(HAVE_MEMCACHED_SASL, 1, [Have SASL support])
       else
-        AC_MSG_ERROR([no, sasl.h is not available. Run configure with --disable-memcached-sasl to disable this check])
+        AC_MSG_ERROR([no, libmemcached built with sasl disabled. Run configure with --disable-memcached-sasl or update libmemcached with sasl support])
       fi
     else
       AC_MSG_RESULT([no])
@@ -298,7 +301,7 @@ if test "$PHP_MEMCACHED" != "no"; then
 
     ORIG_CFLAGS="$CFLAGS"
     ORIG_LIBS="$LIBS"
-  
+
     CFLAGS="$CFLAGS $PHP_LIBMEMCACHED_INCLUDES"
     LIBS="$LIBS $PHP_LIBMEMCACHED_LIBS"
 

